### PR TITLE
Add np.float32 instances validation for FITS headers values.

### LINKED ISSUE #11905: FITS header NaN check does not work for np.float32('nan')?
### Expected behavior

Usually when one tries to set the value of a FITS header key to something like `np.nan`, a `ValueError` will be raised as follows:
```
In [1]: hdr = fits.Header()

In [2]: hdr['test'] = np.nan
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-3bf5b5913e1a> in <module>
----> 1 hdr['test'] = np.nan

~/astro/anaconda3/envs/py3.7/lib/python3.7/site-packages/astropy/io/fits/header.py in __setitem__(self, key, value)
    205             # If we get an IndexError that should be raised; we don't allow
    206             # assignment to non-existing indices
--> 207             self._update((key, value, comment))
    208
    209     def __delitem__(self, key):

~/astro/anaconda3/envs/py3.7/lib/python3.7/site-packages/astropy/io/fits/header.py in _update(self, card)
   1669         else:
   1670             # A new keyword! self.append() will handle updating _modified
-> 1671             self.append(card)
   1672
   1673     def _cardindex(self, key):

~/astro/anaconda3/envs/py3.7/lib/python3.7/site-packages/astropy/io/fits/header.py in append(self, card, useblanks, bottom, end)
   1201             card = Card(card)
   1202         elif isinstance(card, tuple):
-> 1203             card = Card(*card)
   1204         elif card is None:
   1205             card = Card()

~/astro/anaconda3/envs/py3.7/lib/python3.7/site-packages/astropy/io/fits/card.py in __init__(self, keyword, value, comment, **kwargs)
    192                 self.keyword = keyword
    193             if value is not None:
--> 194                 self.value = value
    195
    196         if comment is not None:

~/astro/anaconda3/envs/py3.7/lib/python3.7/site-packages/astropy/io/fits/card.py in value(self, value)
    324         if isinstance(value, float) and (np.isnan(value) or np.isinf(value)):
    325             raise ValueError("Floating point {!r} values are not allowed "
--> 326                              "in FITS headers.".format(value))
    327
    328         elif isinstance(value, str):

ValueError: Floating point nan values are not allowed in FITS headers.
```

I would expect this to be the case for `np.float32('nan')` and its other variants as well.

### Actual behavior

```
In [3]: hdr['test'] = np.float32('nan')

In [4]: hdr
Out[4]: TEST    =                NAN.0
```

### System Details

Linux-2.6.32-754.33.1.el6.x86_64-x86_64-with-centos-6.10-Final
Python 3.7.10 | packaged by conda-forge | (default, Feb 19 2021, 16:07:37)
[GCC 9.3.0]
Numpy 1.21.0
astropy 4.2.1
Scipy 1.7.0
Matplotlib 3.4.2

### PR DESCRIPTION
**Pull request by JasonS09**
### Description

Fixes #11905 
I added a new isinstance() validation for np.float32 values. Program should not allow assigning np.float32('nan') values for FITS headers anymore, instead, a `ValueError` should be raised as with np.float64 values.
Changes has been tested manually, but no automated tests have been written for this issue.
This is my first contribution. Not really sure if there are more optimal approaches to solve the issue. Let me know if I missed something!

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/io/fits/card.py
Comment (saimn) [inline astropy/io/fits/card.py:328]: Please use `isinstance(value, numbers.Real)` instead.

**Review by saimn — CHANGES_REQUESTED** [review]
Thanks for the PR @JasonS09. Can you add a test (in `astropy/io/fits/tests/test_header.py:test_assign_inf_nan`) ?

File: astropy/io/fits/card.py
Comment (JasonS09) [reply astropy/io/fits/card.py]: Done. Added the test as well and tested again.

File: astropy/io/fits/card.py
Comment (saimn) [inline astropy/io/fits/card.py:325]: `numbers.Real` includes `float` so you need only one `isinstance` check.

File: astropy/io/fits/card.py
Comment (saimn) [reply astropy/io/fits/card.py]: Add `import numbers` at the top (https://docs.python.org/3/library/numbers.html).

File: astropy/io/fits/card.py
Comment (JasonS09) [reply astropy/io/fits/card.py:325]: Done. But there is one issue. It appears one test case is failling for a `long_int` negative value, raising an unexpected `TypeError`. `np.isnan()` is not supported for the value, it seems.



 Maybe I should filter out `int` values for the validation? Tested it that way and all test cases worked. Working code looks like this:
```
if (isinstance(value, numbers.Real) and not
    isinstance(value, int)) and (np.isnan(value) or
                                 np.isinf(value)):
```

I'll be waiting for your approval.

Comment (saimn) [reply astropy/io/fits/card.py:325]: Oh right, `numbers.Real` also catches integer which is not what we want here. Then I guess the best is to go back to `isinstance(value, (float, np.float32))`, sorry for the wrong direction.

**Comment by saimn** [comment]
@JasonS09 - Could you squash the commits please ?

File: docs/changes/io.fits/11922.bugfix.rst
Comment (embray) [inline docs/changes/io.fits/11922.bugfix.rst:1]: Please rewrite the second sentence as:

    A ``ValueError`` is now raised if an ``np.float32`` value is assigned to a header.

**Review by embray — CHANGES_REQUESTED** [review]
Just some minor text nitpicks in the changelog entry.  Please fix and then [squash your commits](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#how-to-squash).  Thanks for the fix!

File: docs/changes/io.fits/11922.bugfix.rst
Comment (saimn) [reply docs/changes/io.fits/11922.bugfix.rst:1]: Should be `an ``np.float32`` NaN/Inf` right ?

Comment (embray) [reply docs/changes/io.fits/11922.bugfix.rst:1]: Right, thank you.

File: docs/changes/io.fits/11922.bugfix.rst
Comment (saimn) [inline docs/changes/io.fits/11922.bugfix.rst:1]: I think the first sentence is not useful and can be confusing since the change only concerns NaN/Inf. Suggestion:
```suggestion
Raise ``ValueError`` if an ``np.float32`` NaN/Inf value is assigned to a header keyword.
```

**Review by saimn — APPROVED** [review]
All good, thanks @JasonS09 !

File: docs/changes/io.fits/11922.bugfix.rst
Comment (embray) [reply docs/changes/io.fits/11922.bugfix.rst]: I went back-and-forth on my feelings about that, but I'm inclined to agree.

**Pull request merged**
